### ISSUE TITLE
Overriding `StreamingDataset.__iter__`

### ISSUE THREAD

**Issue by philgzl** [Lightning-AI/litData#552]
I am working with audio files with variable duration. I would like to process the files on-the-fly to yield audio segments of fixed duration `length`. I would like to avoid resorting to looping the files when they are smaller than `length` and subsampling the files when they are longer than `length`.

What I have successfully done in the past is to put files in a queue as they are streamed. If the queue length exceeds `length`, then audio segments are yielded until the queue is shorter than `length`. Else, the next file is loaded and queued. The yielded items can thus be seen as segments sampled from the concatenation of all files in the dataset.

The number of items in the dataset thus depends on `length` and is unknown until the dataset is iterated once. This was not an issue for me as one can iterate over `IterableDataset` instances of unknown length.

Can this be achieved using `StreamingDataset` and `StreamingDataLoader`? I have tried overriding `__iter__` but I faced infinite recursion errors.

**Comment by bhimrazy** [comment]
Hi @philgzl,
I think overriding the `__next__` method might work better than the `__iter__` method, since that's where the actual data is being collected.  
> I’ll give it a try.

By the way, @philgzl, how is your data actually stored? Is it optimized and saved as a 1D array or something similar?  
If so, then perhaps `TokensLoader` might also come in handy.

**Comment by bhimrazy** [comment]
<details><summary> Overriding `__next__` method</summary>

```python
import random
import numpy as np
from tqdm import tqdm
import torch

from litdata import StreamingDataset, StreamingDataLoader, optimize

required_segment_length = 2000  # Target number of audio samples per segment


class AudioDataset(StreamingDataset):
    def __iter__(self):
        self._audio_buffer = []
        super().__iter__()
        return self

    def __next__(self):
        # Fill buffer until we have enough samples to make a full segment
        while len(self._audio_buffer) < required_segment_length:
            try:
                audio_chunk = super().__next__()  # Get next audio tensor
                self._audio_buffer.extend(audio_chunk.tolist())  # Add to buffer
            except StopIteration:
                break  # No more data to read

        if len(self._audio_buffer) < required_segment_length:
            raise StopIteration  # Not enough for a full segment

        # Slice out the segment from the buffer
        segment = self._audio_buffer[:required_segment_length]
        self._audio_buffer = self._audio_buffer[required_segment_length:]  # Keep leftovers

        return torch.tensor(segment, dtype=torch.float32)


def optimize_audio(index):
    # Simulate a variable-length audio file
    audio = np.random.randn(random.randint(1000, 8000)).astype(np.float32)

    # Normalize and convert to int16
    audio /= np.max(np.abs(audio)) if np.max(np.abs(audio)) > 0 else 1.0
    audio = (audio * 32767).astype(np.int16)

    return torch.from_numpy(audio)


if __name__ == "__main__":
    print("Optimizing and writing dataset to disk...")
    optimize(
        fn=optimize_audio,
        inputs=list(range(10)),  # Simulate 10 audio files
        num_workers=4,
        chunk_bytes="64MB",
        output_dir="fast_data",
    )

    print("Loading streaming dataset with fixed-length audio segments...")
    dataset = AudioDataset("fast_data")
    dataloader = StreamingDataLoader(dataset, batch_size=4, num_workers=4, drop_last=True)

    for i, batch in enumerate(tqdm(dataloader)):
        print(f"Batch {i} - shape: {batch.shape}")

```
</details> 

fyi @philgzl — not sure if this is exactly what you were looking for, but let me know if I misunderstood anything or if you have any questions. Thanks!

**Comment by philgzl** [comment]
> how is your data actually stored?

The data is stored as compressed audio in raw bytes. I am decoding the bytes using `soundfile` upon streaming. I suppose this is much more efficient compared to storing uncompressed arrays.

> not sure if this is exactly what you were looking for

Yes, this is the logic I am looking for. I was about to try exactly that but you were faster than me :) will give that a try

Should the `__length__` of the `StreamingDataset` child be updated?

**Comment by bhimrazy** [comment]
> Should the __length__ of the StreamingDataset child be updated?

I don't think so, if you're just using it as an iterator — but you can update it incase needed.

**Comment by philgzl** [comment]
Okay it seems like this works! I overrode `__len__` to return `None` in both my dataset and my dataloader to prevent displaying misleading information when wrapping with `tqdm`.

Thanks!

**Comment by tchaton** [comment]
Hey @philgzl. Would you mind to create & publish a Studio to demonstrate the advantage of litdata for audio vs not using it.

I would add it to the docs, so it becomes a reference.

**Comment by philgzl** [comment]
I can give it a try
